src/transformers/models/whisper/modeling_whisper.py

```diff
@@ -1519,18 +1525,37 @@ def generate(
 
         if hasattr(generation_config, "is_multilingual") and generation_config.is_multilingual:
             if hasattr(generation_config, "language"):
-                forced_decoder_ids.append((1, generation_config.lang_to_id[generation_config.language]))
+                if generation_config.language in generation_config.lang_to_id.keys():
+                    language_token = generation_config.language
+                elif generation_config.language in TO_LANGUAGE_CODE.keys():
+                    language_token = f"<|{TO_LANGUAGE_CODE[generation_config.language]}|>"
+                else:
+                    raise ValueError(f"The `{generation_config.language}` language token is not supported.")
+                forced_decoder_ids.append((1, generation_config.lang_to_id[language_token]))
             else:
                 forced_decoder_ids.append((1, None))
 
             if hasattr(generation_config, "task"):
-                forced_decoder_ids.append((2, generation_config.task_to_id[generation_config.task]))
+                if generation_config.task in TASK_IDS:
+                    forced_decoder_ids.append((2, generation_config.task_to_id[generation_config.task]))
+                else:
+                    raise ValueError(
+                        f"The `{generation_config.task}`task is not supported. The task should be one of `{TASK_IDS}`"
+                    )
             else:
                 forced_decoder_ids.append((2, generation_config.task_to_id["transcribe"]))
 
-        if (
-            hasattr(generation_config, "return_timestamps") and generation_config.return_timestamps
-        ) or return_timestamps:
+        # Legacy code for backward compatibility
+        if hasattr(self.config, "forced_decoder_ids") and forced_decoder_ids != self.generation.forced_decoder_ids:
```

Comment:
Should this be `generation_config.forced_decoder_ids`? We don't have an attribute `self.generation` AFAIK
```suggestion
        if hasattr(self.config, "forced_decoder_ids") and forced_decoder_ids != generation_config.forced_decoder_ids:
```

I think we should favour `forced_decoder_ids` created using the `generation_config` over the `config`. Currently, as things stand, if a user passes `language=..., task=...`, these arguments will be ignored, and the forced decoder ids will always be set based on `config.forced_decoder_ids`.

What I think we should do is only set `forced_decoder_ids = config.forced_decoder_ids` **if** neither `language` or `task` are passed.

This way, a user can still control the forced decoder ids using the generate args even if they've set `config.forced_decoder_ids`. This means older Whisper models will still be compatible with the new generate method.

WDYT?

Reply:
Agreed! Thanks for the suggestion